### PR TITLE
[WIP] Multilabel Detection

### PR DESCRIPTION
**Pull request by hadware**
This is a new PR for the VTC feature, this time based on a cleaner implem. I'm making a new PR as to keep the former branch "clean" (and prevent any mishaps).

What is done:
- renaming the `SpeakerTracking` task into a `MultilabelDetection` task
- added `MultilabelPipeline`
- update `MultilabelFscore.report()` 
- tested the new preprocessor

What's to be done:
- [ ] re-test the new implem on our clinical data (as well as the child data from @MarvinLvn )
- [ ] maybe a couple of unit tests (especially for the preprocessor)
- [ ] maybe make the aggregated "multilabel" fscore duration-based instead of file-based

### CODE REVIEW

File: pyannote/audio/tasks/segmentation/voice_type_classification.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            warm_up=warm_up,
+            batch_size=batch_size,
+            num_workers=num_workers,
+            pin_memory=pin_memory,
+            augmentation=augmentation,
+        )
+        self.balance = balance
+        self.weight = weight
+
+        self.clsf_specs = MultilabelDetectionSpecifications. \
```
Comment (hbredin) [inline pyannote/audio/tasks/segmentation/voice_type_classification.py:66]: Before I go further in the review...
Didn't we agree that this should be done by a preprocessor? 
Shouldn't you remove this VTC task completely and have this PR focus on MultilabelDetection?

Comment (hadware) [reply pyannote/audio/tasks/segmentation/voice_type_classification.py]: Yup! That was a mistake on my part, I removed the file. Everything is in `multilabel_detection` now (both for pipelines and task)

**Comment by hbredin** [comment]
Seems to go in the right direction.
I will make a pass and update a few things myself.

In the meantime, could you add a notebook showing how to use this new objects to train a male/female/speech multi-label pipeline on top of `Debug.SpeakerDiarization.Debug` protocol available in `tests/data/database.yml`.

Basically, this would illustrate
* the use of a preprocessor to convert `Mxxxx` labels to `male` and `Fxxxx` labels to `female`
* the use of a VoiceTypeClassifierPreprocessor preprocessor to add the union of `male` and `female` as `speech`
* [a way to chain them](https://github.com/pyannote/pyannote-database/issues/80)
* the new `MultilabelDetection` task
* the new `MultilabelDetection`pipeline

**Comment by hbredin** [comment]
I pushed a few changes and might have broken things... Looking forward to the notebook mentioned above to see if things still work :)

**Comment by hadware** [comment]
Alright, will do (and iirc i've already started writing the notebook, it shouldn't be long). regarding the chaining thing, i've made an attempt [here](https://github.com/hadware/pyannote-vtc-testing/blob/a541bf8945b991a8f7e25174f40c4ed985440ef0/main.py#L28).... but i'm not sure this is the best way to do it. I'm open to suggestions.

**Comment by hadware** [comment]
(and thanks a lot for your changes btw :smiley: )

**Comment by hadware** [comment]
"EDIT": sorry, i realized the code sample I linked was the same as the one in the issue you referenced... I thought I had tweaked some things.

**Comment by hadware** [comment]
Alright,

I edited the sketch I had for a VTC tutorial (renamed it to multilabeling). I still used "VTC" as a task name throughout the tutorial, don't know how you feel about that.  I tried to do my best to show how to use the `Debug.SpeakerDiarization.Debug` db for testing. Note that I also have a VTC debug database [here](https://github.com/hadware/pyannote-db-vtc-debug) (complete with fake audio and all). It's a bit more "complete" as it has 3 base classes and 2 union classes.

I still need to run the full tutorial on a machine with GPU's as to get the right cells outputs.

Regarding the `DeriveMetaLabels` preproc, I wondered about some tiny yet tricky details: should we allow for some unions/intersections to be based on classes not contained in the `classes` parameter? (this is currently "defensively" enforced using an `assert`)

EG: annotation contains "MALE", "FEMALE", and "CHILD". We'd like the model to detect "MALE", "FEMALE", and "SPEECH" = {"MALE", "FEMALE", "CHILD"}

**Comment by hbredin** [comment]
> Regarding the `DeriveMetaLabels` preproc, I wondered about some tiny yet tricky details: should we allow for some unions/intersections to be based on classes not contained in the `classes` parameter? (this is currently "defensively" enforced using an `assert`)

I think this use case should be allowed.

**Comment by hadware** [comment]
Right!

I'll change that azappe.

**Comment by hadware** [comment]
Done.

**Comment by hbredin** [comment]
> I still need to run the full tutorial on a machine with GPU's as to get the right cells outputs.

Any update on this? 

Also -- the API has changed a bit recently and your PR is most likely impacted.

**Comment by hadware** [comment]
Well, merged, updated the tutorial, and tested it. It works (in the sense that the code runs fine). However, the dataset is a bit too "debuggy" for this task and it performs extremely poorly (bug the graphs are nice... I guess?).

A new engineer in our lab is going to test the task on the infamous BabyTrain dataset next week.

**Comment by hadware** [comment]
:warning: Important note: **this requires a new version (containing the parameter dicts) of pyannote-pipeline to be released, and the requirements for this package to be bumped to this version**

**Comment by hbredin** [comment]
FYI, I did not forget about this PR and will eventually have a look at it.

I think I might update the notebook to rely on the full AMI so that the resulting model doesn't seem completely useless :)

Also, I would like to have this notebook runnable on Google Colab so I started writing one that takes care of downloading and setting up AMI protocol. What do you think?
https://colab.research.google.com/drive/1zQLaZkJrdsIgxEBk6TIp48vrM79vC4Qj?usp=sharing

This might also serve as an easy way for users to share minimum reproducible example in their bug reports (which are usually not very useful right now...)

**Comment by hadware** [comment]
Ok ok, I'm at the ready for a review :)

I think a google colab notebook would be a great idea ! Should I copy and adapt your AMI notebook for segmentation?

**Comment by hbredin** [comment]
> I think a google colab notebook would be a great idea ! Should I copy and adapt your AMI notebook for segmentation?

I did not dare to ask you :) That would be awesome!

**Comment by hadware** [comment]
Just so you know, i'm working on the Colab notebook, it's just that the training is taking a bit longer than expected. It's [here](https://colab.research.google.com/drive/1JX3x4t_QTHMr8hzjw3Em8eQBQnQ3UBgk?usp=sharing) if you want to check it out. I had to tweak a bit the AMI loading part because some things went wrong with your excerpt

File: pyannote/audio/pipelines/multilabel_detection.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                label: ParamDict(
+                    onset=Uniform(0.0, 1.0),
+                    offset=Uniform(0.0, 1.0),
+                    min_duration_on=Uniform(0.0, 2.0),
+                    min_duration_off=Uniform(0.0, 2.0),
+                )
+                for label in self._classes
+            }
+        )
+        # TODO: would it make sense to share min_duration_{on|off} between classes?
```
Comment (hbredin) [inline pyannote/audio/pipelines/multilabel_detection.py:106]: It would. We could add a `share_min_duration` flag to `__init__`

[the remark this answers is not in the document]
Comment (hbredin) [reply pyannote/audio/pipelines/multilabel_detection.py:106]: For this particular task that may lead to a huge hyperparameter search space, we could also force `onset == offset` (once again through a flag `use_hysteresis`).

[the remark this answers is not in the document]
Comment (hbredin) [reply pyannote/audio/pipelines/multilabel_detection.py:106]: But, let's do that in a separate PR.

File: pyannote/audio/tasks/segmentation/multilabel_detection.py [hunk 3]
```diff
@@ -78,19 +80,19 @@ class SpeakerTracking(SegmentationTaskMixin, Task):
...
+            classes: Optional[List[str]] = None,
+            duration: float = 2.0,
+            warm_up: Union[float, Tuple[float, float]] = 0.0,
+            balance: Text = None,
+            weight: Text = None,
+            batch_size: int = 32,
+            num_workers: int = None,
+            pin_memory: bool = False,
+            augmentation: BaseWaveformTransform = None,
+            metric: Union[Metric, Sequence[Metric], Dict[str, Metric]] = None,
```
Comment (hbredin) [inline pyannote/audio/tasks/segmentation/multilabel_detection.py:94]: I haven't thought a lot about the validation metric for this type of task.
It relies on AUROC right? But what does it mean for multilabel classification?

Comment (hbredin) [inline pyannote/audio/tasks/segmentation/multilabel_detection.py:94]: I haven't thought a lot about the validation metric for this type of task.
It relies on AUROC right? But what does it mean for multilabel classification?

**Review by hbredin — COMMENTED** [review]
I spent some time looking at the code. Looks good!
I do have a few questions and would love an update on the notebook :) Does it run as it is right now (refering to the link you shared earlier)?

**Comment by hbredin** [comment]
FYI - I just release pyannote.pipeline 2.3 with support for `ParamDict`py.

**Comment by hadware** [comment]
All _right_

https://colab.research.google.com/drive/1JX3x4t_QTHMr8hzjw3Em8eQBQnQ3UBgk#scrollTo=U9lwdYcTHyp_ is working fine as an example (with your input on using a subset of AMI). The inferred annotations look pretty good, and the tuning step amounts to a noticeable gain of bout 5% of IER.

If you think this is good enough, I can then re-add the comments and tutorial text from the original notebook (with a couple of tweaks) and we can call it a day. I'll also merge the latest commits from "upstream".

**Comment by hadware** [comment]
I'll also take care of the comments you submitted in https://github.com/pyannote/pyannote-audio/pull/891#pullrequestreview-1000345023

File: pyannote/audio/tasks/segmentation/multilabel_detection.py (hunk 3 above)
Comment (hadware) [reply pyannote/audio/tasks/segmentation/multilabel_detection.py:94]: It does rely on AUROC I think. I don't have an answer to this. I just used the metric provided by the class without thinking a second about what I was doing. :upside_down_face:

**Comment by hbredin** [comment]
> All _right_
> 
> https://colab.research.google.com/drive/1JX3x4t_QTHMr8hzjw3Em8eQBQnQ3UBgk#scrollTo=U9lwdYcTHyp_ is working fine as an example (with your input on using a subset of AMI). The inferred annotations look pretty good, and the tuning step amounts to a noticeable gain of bout 5% of IER.

Awesome! I was able to run the whole thing successfully in 10min or so. 
That is a great demo! 

> If you think this is good enough, I can then re-add the comments and tutorial text from the original notebook (with a couple of tweaks) and we can call it a day. I'll also merge the latest commits from "upstream".

Yes, please!

**Comment by hadware** [comment]
I re-added the text to the colab notebook, and added, implemented and tested the `share_min_duration` option

**Comment by hbredin** [comment]
Hey @hadware, would you mind adding the notebook to this PR?

**Comment by hadware** [comment]
Done.

**Comment by hbredin** [comment]
Hurrah! Looking forward to lots of new applications!

I removed the notebook because it felt like it was not clean enough.
Anyway, all notebooks should probably be updated to run on Google Colab with the new AMI Mini.
I'll open a dedicated issue.

Thanks again @hadware for the hard work and sorry for being such a pinailleur...

**Comment by hadware** [comment]
The pinailling is part of the fun :hot_face: 

I'd like to pretrain a couple of models to host then on hugginface along with the others. What kind of classes/train datasets would you suggest? I was thinking about `MALE`/`FEMALE` using AMI/VoxCeleb, but that's it. Any other ideas?
